app/extract_plugin_tar.go

```diff
@@ -72,8 +72,12 @@ func extractTarGz(gzipStream io.Reader, dst string) error {
 			if err != nil {
 				return err
 			}
-			defer outFile.Close()
-			if _, err := io.Copy(outFile, tarReader); err != nil {
+			_, err = io.Copy(outFile, tarReader)
```

Comment:
The `defer` statements would stack up without actually closing the files until the extract was finished.

Reply:
In cases like this, I usually wrap the code in an anonymous function and use defer as usual there. So something like:

```go
copyFile := func(path string, reader *tar.Reader) error {
	outFile, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(header.Mode))
	if err != nil {
		return err
	}
	defer outFile.Close()

	if _, err := io.Copy(outFile, reader); err != nil {
		return err
	}
}

if err := copyFile(path, tarReader); err != nil {
	return err
}
```

0/5

Reply:
Having `defer` statements in loops are always tricky. I think using an anonymous function like @agnivade suggested would be great here to call `defer` immediately and also not do `io.Copy` in case of an error.

Reply:
Great suggestions, @agnivade & @isacikgoz! For the wrapping function, I opted just to directly access the variables from the existing closure instead of passing them into the new `copyFile`, while still enabling the `defer`. Any concerns with that style?

Reply:
No concerns from my side. Looks great.